### PR TITLE
Add API docs for #tap and #dup

### PR DESCRIPTION
**Pull request by hachi8833**
Retry of #792.

### CODE REVIEW

File: vm/class.go [hunk 1]
```diff
@@ -870,6 +870,33 @@ var builtinClassCommonInstanceMethods = []*BuiltinMethodObject{
 		},
 	},
 	{
+		// Performs a 'shallow' copy of the receiver object and returns it.
+		// Any arguments are just ignored.
+		// The object_id of the returned object is different from the one of the receiver.
+		// Note that `#tap` just returns self without copying if the receiver is an immutable object
```
Comment (st0012) [inline vm/class.go]: Do we need to explain `tap` here?

File: vm/class.go [hunk 2]
```diff
@@ -1545,6 +1572,38 @@ var builtinClassCommonInstanceMethods = []*BuiltinMethodObject{
 
 		},
 	},
+	// Just evaluates a given block with the receiver and returns the receiver.
+	// #tap method literally "taps into" the method chain and
+	// good for inspecting method chains.
+	// Any arguments to the method are just ignored.
+	//
+	// ```ruby
+	// (1..10).tap do |x|
```
Comment (st0012) [inline vm/class.go]: can also put

```ruby
a = (1..10)
a.tap do |x|
# omit the rest
puts(a) #=> (1..10)
```

to show the original object is not touched

File: vm/hash.go [hunk 3]
```diff
@@ -426,6 +426,45 @@ var builtinHashInstanceMethods = []*BuiltinMethodObject{
 		},
 	},
 	{
+		// Performs a 'shallow' copy of the hash and returns it.
+		// Any arguments are ignored.
+		// The object_id of the returned object is different from the one of the receiver.
+
+		// Caveat: any keys of hash ARE also copied with different object ids for now.
```
Comment (st0012) [inline vm/hash.go]: This shouldn't be the case, @hachi8833 can you open an issue for this and I'll fix it?

File: vm/class.go (hunk 1 above)
Comment (hachi8833) [reply vm/class.go]: Redundant as pointed out 😓.

File: vm/class.go (hunk 2 above)
Comment (hachi8833) [reply vm/class.go]: Got it

File: vm/hash.go (hunk 3 above)
Comment (hachi8833) [reply vm/hash.go]: OK!

Comment (st0012) [reply vm/hash.go]: This won't be solved in the near future. Can you update other comments and I'll merge it?

Comment (hachi8833) [reply vm/hash.go]: @st0012 how about this?

> 		// Performs a copy of the hash, including the keys and values, and returns it.
>		// Any arguments are ignored.
>		// The object_id of the returned object is different from the one of the receiver.
>		// Caveat: any keys of hash ARE also copied with different object ids for now.
>		// This comes from the fact that the string objects are NOT frozen in current Goby.
>		// TODO: Perhaps the `Hash#dup` will not copy keys when we change the implementation of string objects as frozen.

Comment (st0012) [reply vm/hash.go]: Think the TODO part is unnecessary here. Otherwise looks good.

Comment (hachi8833) [reply vm/hash.go]: Thnx!

**Comment by st0012** [comment]
@hachi8833 thanks!

**Pull request merged**